driver: sensor: tmp114: Add support for frequency attribute

fgihl opened the pull request:
Support setting sample frequency via attr_set and dts.
Also support setting oversample attr via dts.

fgihl commented:
Will wait until tmp116 is merged and then try to make this similar.

fgihl commented:
I would like to have the stale label removed.

fgihl commented:
These changes are based on the changes in tmp116 (#77263), therefore I wanted tmp116 to approve first.
So the only dependency is that I wanted the two driver to have similar implementation.

jeppenodgaard commented:
Sorry about the wait @fgihl. I've been absent for a while. I just updated the PR this depends on based on @ubieda's suggestion.

fgihl commented:
I tried to align this PR according to #77263.

jeppenodgaard commented on drivers/sensor/ti/tmp114/tmp114.c:
```diff
@@ -36,6 +41,8 @@ struct tmp114_data {
 struct tmp114_dev_config {
 	struct i2c_dt_spec bus;
+	uint16_t odr;
+	int oversampling;
```
Why not boolean type to match dts property value?

jeppenodgaard commented on drivers/sensor/ti/tmp114/tmp114.c:
```diff
@@ -161,11 +191,47 @@ static int tmp114_attr_get(const struct device *dev, enum sensor_channel chan,
-	int16_t value;
-	int rc;
+	int16_t avg = 0;
+	uint16_t conv = 0;
```
A shared generic `value` more is more scalable if/when more cases are added to the switch. I also think it aligns with [Linux naming coding style](https://www.kernel.org/doc/html/latest/process/coding-style.html#naming).
Changing it from `int16_t` to `uint16_t` seems correct.

jeppenodgaard reviewed (changes requested):
Thank you for doing clang-format in a separate commit!

Please split the last commit into two commits.
 - dts and `atts_set` sample frequency support.
 - dts oversampling support.

Try to avoid changing more than necessary for each commit. E.g. renaming, refactoring.

MaureenHelm commented on drivers/sensor/ti/tmp114/tmp114.c:
```diff
@@ -197,12 +191,10 @@ static int tmp114_attr_set(const struct device *dev,
-	.sample_fetch = tmp114_sample_fetch,
-	.channel_get = tmp114_channel_get
-};
+static const struct sensor_driver_api tmp114_driver_api = {.attr_get = tmp114_attr_get,
```
Need to restore usage of `DEVICE_API` here

fgihl replied:
I think DEVICE_API macro is the preferred way?
I have move this change to a separate commit.

MaureenHelm replied:
Yes DEVICE_API is the preferred way. It was correct before this commit, please amend the commit to not remove the DEVICE_API usage.

fgihl replied:
Sorry, I got lost in the different commit.  Did not realise I remove it here.
Thanks for pointing it out.

MaureenHelm commented on drivers/sensor/ti/tmp114/tmp114.c:
```diff
@@ -9,25 +9,30 @@
+#define TMP114_REG_ALERT     0x2
+#define TMP114_REG_CFGR      0x3
+#define TMP114_REG_DEVICE_ID 0xb
 
-#define TMP114_RESOLUTION	78125       /* in tens of uCelsius*/
-#define TMP114_RESOLUTION_DIV	10000000
+#define TMP114_RESOLUTION     78125 /* in tens of uCelsius*/
+#define TMP114_RESOLUTION_DIV 10000000
 
-#define TMP114_DEVICE_ID	0x1114
+#define TMP114_DEVICE_ID 0x1114
 
-#define TMP114_ALERT_DATA_READY  BIT(0)
-#define TMP114_AVG_MASK          BIT(7)
+#define TMP114_ALERT_DATA_READY BIT(0)
```
whitespace refactoring looks unrelated to setting the ODR

fgihl replied:
I change and tried to minimise the whitespace changes, but TMP114_AVG and TMP114_ALERT_DATA_READY needed to be move for the code to look ok.

MaureenHelm commented on drivers/sensor/ti/tmp114/tmp114.c:
```diff
@@ -213,19 +282,25 @@ static int tmp114_init(const struct device *dev)
 	if (rc < 0) {
 		return rc;
 	}
-	LOG_INF("Got device ID: %x", id);
+	LOG_DBG("Got device ID: %x", id);
```
Unrelated to setting the ODR

MaureenHelm commented on drivers/sensor/ti/tmp114/tmp114.c:
```diff
@@ -213,19 +282,25 @@ static int tmp114_init(const struct device *dev)
 	if (rc < 0) {
 		return rc;
 	}
-	LOG_INF("Got device ID: %x", id);
+	LOG_DBG("Got device ID: %x", id);
 	drv_data->id = id;
 
+	rc = tmp114_write_config(dev, TMP114_CFGR_CONV, cfg->odr);
+	if (rc < 0) {
+		return rc;
+	}
+
 	return 0;
 }
 
 #define DEFINE_TMP114(_num) \
 	static struct tmp114_data tmp114_data_##_num; \
 	static const struct tmp114_dev_config tmp114_config_##_num = { \
-		.bus = I2C_DT_SPEC_INST_GET(_num) \
+		.bus = I2C_DT_SPEC_INST_GET(_num), \
+		.odr = DT_INST_PROP(_num, odr), \
 	}; \
-	SENSOR_DEVICE_DT_INST_DEFINE(_num, tmp114_init, NULL, \
-		&tmp114_data_##_num, &tmp114_config_##_num, POST_KERNEL, \
+	SENSOR_DEVICE_DT_INST_DEFINE(_num, tmp114_init, NULL, &tmp114_data_##_num, \
+		&tmp114_config_##_num, POST_KERNEL, \
```
Unrelated to setting the ODR

MaureenHelm commented on dts/bindings/sensor/ti,tmp114.yaml:
```diff
@@ -6,3 +6,20 @@ description: Texas Instruments TMP114 temperature sensor
+      Specify the default temperature output data rate in milliseconds (ms).
+      Default is power-up configuration (250ms).
+    enum:
+      - 0    # TMP114_DT_ODR_6_4_MS
+      - 1    # TMP114_DT_ODR_32_25_MS
+      - 2    # TMP114_DT_ODR_62_5_MS
+      - 3    # TMP114_DT_ODR_125_MS
+      - 4    # TMP114_DT_ODR_250_MS
+      - 5    # TMP114_DT_ODR_500_MS
+      - 6    # TMP114_DT_ODR_1000_MS
+      - 7    # TMP114_DT_ODR_2000_MS
```
Move the enum descriptions into the description property so they get rendered in the documentation
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules

fgihl committed:
drivers: sensor: tmp114: device tree support for setting oversample

Add support for setting oversample from `attr_set` and device tree

Signed-off-by: Fredrik Gihl <fgihl@hotmail.com>

fgihl committed:
drivers: sensor: tmp114: Correct clang warning

Corrected clang warning introduced on commit 9eaf488

Signed-off-by: Fredrik Gihl <fgihl@hotmail.com>

fgihl commented on drivers/sensor/ti/tmp114/tmp114.c:
```diff
@@ -165,13 +189,46 @@ static int tmp114_attr_get(const struct device *dev, enum sensor_channel chan,
+	const uint32_t freq_micro = sensor_value_to_micro(frequency);
+
+	switch (freq_micro) {
+	case 156000000:
```
As you say, I think that's out of the scope for this PR.
I can change tmp116 and this driver later as a separate PR.

yperess replied:
This is not out of scope for this PR, please update this to round to the best frequency

```c
if (freq_micro >= 156000000) {
  *config_value = TMP114_DT_ODR_6_4_MS; /* 6.4 ms */
} else if (freq_micro >= 32000000) {
  *config_value = TMP114_DT_ODR_31_25_MS; /* 32.25 ms */
} else ...
```

jeppenodgaard replied:
I created https://github.com/zephyrproject-rtos/zephyr/issues/86034.

fgihl replied:
@jeppenodgaard : I updated the driver to return `-ENOTSUP` as stated in #86034.

dipakgmx commented on include/zephyr/dt-bindings/sensor/tmp114.h:
Since you have added ODR values in the bindings file([ti,tmp114.yaml](https://github.com/zephyrproject-rtos/zephyr/pull/77248/files#diff-4e52f357354ac5bce95ebd082585ad70c19baf3b7ffb4eda68e66575ca626e5c), this file is not needed anymore.

fgihl replied:
sdfsdfsd

dipakgmx commented on drivers/sensor/ti/tmp114/tmp114.c:
```diff
@@ -224,13 +276,28 @@ static int tmp114_init(const struct device *dev)
+	val.val1 = cfg->oversampling ? 1 : 0;
+	val.val2 = 0;
+
+	rc = tmp114_attr_set(dev, SENSOR_CHAN_AMBIENT_TEMP, SENSOR_ATTR_OVERSAMPLING, &val);
```
Why is the the `sensor_attr_set` function being used in an init function?

fgihl replied:
In the applications device tree the you set the oversampling value (i.e. the frequency the chip should sample with).
Here we take the device tree value and push it down to the chip.

dipakgmx commented on drivers/sensor/ti/tmp114/tmp114.c:
```diff
@@ -165,13 +189,36 @@ static int tmp114_attr_get(const struct device *dev, enum sensor_channel chan,
-			   enum sensor_channel chan,
-			   enum sensor_attribute attr,
-			   const struct sensor_value *val)
+static int tmp114_odr_value(const struct sensor_value *frequency, uint16_t *config_value)
```
Any reason to not call this function in the init function?

fgihl replied:
__tmp114_odr_value__ only convert the frequency value to a an odd value.
The application sets frequency when calling __sensor_attr_set(dev, SENSOR_CHAN_AMBIENT_TEMP, SENSOR_ATTR_SAMPLING_FREQUENCY, &sample_freq);__

yperess commented on drivers/sensor/ti/tmp114/tmp114.c:
```diff
@@ -165,13 +189,36 @@ static int tmp114_attr_get(const struct device *dev, enum sensor_channel chan,
-	int rc;
+	const uint32_t freq_micro = sensor_value_to_micro(frequency);
+
+	if (freq_micro <= 500000) {
```
Is it not possible to set the frequency to 0Hz to turn off the sampling?

fgihl replied:
No. The chip only supports continuous sampling or shutdown mode. In continuous mode, the slowest speed is 2Hz.
The driver only supports continuous mode.

fgihl committed:
drivers: sensor: tmp114: Support setting odr

Add support for setting the sample frequency via `attr_set` and the
output data rate from device tree source.

Signed-off-by: Fredrik Gihl <fgihl@hotmail.com>

fgihl commented:
Updated the driver to return  `-ENOTSUP` if the requested frequence value is too high.
